language-examples/remedial-js/9.0-ES5-vs-ES6/README.md

```diff
@@ -0,0 +1,961 @@
+1. [About](#about)
+1. [License](#license)
+
+## Arrow Functions
+
+An arrow function expression (also known as fat arrow function) has a shorter syntax compared to function expressions and lexically binds the this value. Arrow functions are always anonymous.
+
+
+ES6:
+
+```js
+[1, 2, 3].map(n => n * 2);
+// -> [ 2, 4, 6 ]
+```
+
+ES5 equivalent:
+
+```js
+[1, 2, 3].map(function(n) { return n * 2; }, this);
+// -> [ 2, 4, 6 ]
+```
+
+ES6:
+
+```js
+var evens = [2, 4, 6, 8, 10];
+
+// Expression bodies
+var odds = evens.map(v => v + 1);
+var nums = evens.map((v, i) => v + i);
+
+console.log(odds);
+// -> [3, 5, 7, 9, 11]
+
+console.log(nums);
+// -> [2, 5, 8, 11, 14]
+
+// Statement bodies
+var fives = [];
+nums = [1, 2, 5, 15, 25, 32];
+nums.forEach(v => {
+  if (v % 5 === 0)
+    fives.push(v);
+});
+
+console.log(fives);
+// -> [5, 15, 25]
+
+// Lexical this
+var bob = {
+  _name: 'Bob',
+  _friends: [],
+  printFriends() {
+    this._friends.forEach(f =>
+      console.log(this._name + ' knows ' + f));
+  }
+}
+```
+
+ES5:
+
+```js
+'use strict';
+
+var evens = [2, 4, 6, 8, 10];
+
+// Expression bodies
+var odds = evens.map(function (v) {
+  return v + 1;
+}, this);
+var nums = evens.map(function (v, i) {
+  return v + i;
+}, this);
+
+console.log(odds);
+// -> [3, 5, 7, 9, 11]
+
+console.log(nums);
+// -> [2, 5, 8, 11, 14]
+
+var fives = [];
+nums = [1, 2, 5, 15, 25, 32];
+
+// Statement bodies
+nums.forEach(function (v) {
+  if (v % 5 === 0) {
+    fives.push(v);
+  }
+}, this);
+
+console.log(fives);
+// -> [5, 15, 25]
+
+// Lexical this
+var bob = {
+  _name: 'Bob',
+  _friends: [],
+  printFriends: function printFriends() {
+    this._friends.forEach(function (f) {
+      return console.log(this._name + ' knows ' + f);
+    }, this);
+  }
+};
+```
+
+## Block Scoping Functions
+
+Block scoped bindings provide scopes other than the function and top level scope. This ensures your variables don't leak out of the scope they're defined:
+
+ES6:
+
+```js
+// let declares a block scope local variable,
+// optionally initializing it to a value in ES6
+
+'use strict';
+
+var a = 5;
+var b = 10;
+
+if (a === 5) {
+  let a = 4; // The scope is inside the if-block
+  var b = 1; // The scope is inside the function
+
+  console.log(a);  // 4
+  console.log(b);  // 1
+}
+
+console.log(a); // 5
+console.log(b); // 1
+```
+
+ES5:
+
+```js
+'use strict';
+
+var a = 5;
+var b = 10;
+
+if (a === 5) {
+  // technically is more like the following
+  (function () {
+    var a = 4;
+    b = 1;
+
+    console.log(a); // 4
+    console.log(b); // 1
+  })();
+}
+
+console.log(a); // 5
+console.log(b); // 1
+```
+
+
+ES6:
+
+```js
+// const creates a read-only named constant in ES6.
+'use strict';
+// define favorite as a constant and give it the value 7
+const favorite = 7;
+// Attempt to overwrite the constant
+try {
+  favorite = 15;
+} catch (err) {
+  console.log('my favorite number is still: ' + favorite);
+  // will still print 7
+}
+```
```

Comment:
I think it might be worth mentioning that only attempting to bind a new value to a variable declared with `const` throws. it does not protect against mutation. 
```javascript
const myObj = { name: 'constobj' }
myObj.name = 'mutableObj'
myObj.name // 'mutableObj'

myObj = { name: 'wholenewObj' } // error
```